# 1. State Problem

### 1.1 Problem and Pain point
ในปัจจุบันนั้น ประเทศไทยมีจำนวนอุบัติเหตุที่เกิดขึ้นบนท้องถนนเป็นอันดับต้นๆของโลก โดยอัตราส่วนประเภทของพาหนะที่เกิดอุบัติเหตุเป็นรถจักรยานยนต์ถึง 73% ซึ่งในจำนวนนั้นมีผู้เสียชีวิตและบาดเจ็บสาหัสจากการไม่สวมใส่หมวกกันน็อคสูงถึง 45% ซึ่งกลุ่มของเราได้เล็งเห็นปัญหานี้จากบริเวณใกล้ตัวซึ่งก็คือภายในสถาบันจากการสังเกตุพบว่านักศึกษาจำนวนมากที่ใช้รถจักรยานยนต์ในการเดินทางนั้นไม่ได้ส่วมใส่หมวกกันน็อคเป็นจำนวนมากซึ่งทางมหาลัยได้มีการตั้งกฎระเบียบใหม่แต่ก็ยังไม่เป็นผลไม่สามารถลดจำนวนคนที่ไม่ใส่หมวกกันน็อคได้ เราจึงได้พัฒนาระบบตรวจจับคนไม่ใส่หมวกกันน็อคผ่าน Computer vision และทำการตรวจจับใบหน้าเพื่อบันทึกการทำผิดกฎจราจรเพื่อใช้ในการพิจารณาบทลงโทษต่อไป

### 1.2 Related work
- "เชียงใหม่" นำร่องใช้กล้อง AI ตรวจจับคนเมินใส่หมวกกันน็อก [link](https://www.thaipbs.or.th/news/content/298927)
- Helmet Detection System for Motorcycle Riders with Explainable Artificial Intelligence Using Convolutional Neural Network and Grad-CAM [link](https://kukr.lib.ku.ac.th/kukr_es/kukr/search_detail/result/422154)

### 1.3 Unique method
โปรเจ็คของเรามีจุดเด่นแตกต่างกับคนอื่นตรงที่เรามีการทำ report สรุปผลการตรวจจับได้ตามช่วงเวลาที่เราต้องการผ่าน เว็บไซต์ ได้โดยที่ตัว report ของเรามาจากการใช้ large language model ในการ generate report ให้ตามข้อมูลที่ถูกเก็บจากการตรวจพบคนที่ไม่สวมหมวกกันน็อคในแต่ละครั้ง

# 2. Data Preparation

### 2.1 Data source
ข้อมูลที่ทางทีมของเรานำมาใช้นั้นนำมาจากภาพบนอินเตอร์เน็ตรวมกับภาพที่ทีมของเราไปเก็บภาพจากบริเวณรอบสถาบันและสถานที่ที่จะติดตั้งอุปกรณ์จริงเพื่อให้ได้ภาพและชุดข้อมูลที่มีปริมาณมากเพียงพอให้โมเดลมีความแม่นยำรวมไปถึงตอบโจทย์ให้โมเดลมีความแม่นยำในมุมภาพของรถมอเตอร์ไซต์ในพื้นที่จริง รวมไปถึงรูปภาพของหมวกกันน็อคและศรีษะของคน [link to dataset](https://drive.google.com/file/d/1tsFZXPXTrGPLBL8JeAjN-nQJFPezHclR/view?usp=sharing)

### 2.2 Data Cleaning
เนื่องจากทีมของเราได้แบ่งให้ทุกๆคนช่วยกันหาข้อมูลบนอินเตอร์เน็ตซึ่งเกิดปัญหาซ้ำซ้อนของภาพได้เราจึงทำการลบภาพที่ซ้ำกันออกไป 1 ภาพ และเราได้มีการคัดและลบภาพที่เบลอหรือไม่ได้สื่อถึงวัตถุที่เราต้องการตวจจับออกไปเพื่อให้ข้อมูลมีคุณภาพมากขึ้น

### 2.3  การแปลงข้อมูล (Data Transformation)
ในการแปลงข้อมูลนั้นเราได้ใช้เว็บไซต์ Roboflow เขามาช่วยในการใส่และระบุชนิดของวัตถุที่อยู่ในภาพโดยเราได้แบ่งเป็น 3 ชนิดหลักๆคือ รถจักรยานยนต์ หมวกกันน็อค และ หัวของคน หลังจากเราทำการใส่ข้อมูลแล้ว เราจะทำการนำชุดข้อมูลออกมาใช้งานโดยมีการเปลี่ยนสัดส่วนรูปภาพเป็น 640x640 ก่อนนำมาใช้เทรนโมเดล
<center><div>
<img src="img/step_prepro2.png" height="300" style="display=block; margin:auto"/>
<img src="img/step_prepro.png" height="300" style="display=block; margin:auto"/>
</div></center>


# 3. Train

### 3.1 AI Model
โมเดลที่เราเลือกใช้ในการทำ Computer vision classification นั้นเป็น Yolo v8 เนื่องจากตัวของโมเดลนั้นสามารถใช้งานได้ง่ายมีการทำเครือข่ายประสาทเทียมภายในให้เรียบร้อยเราเพียงแค่นำข้อมูลภาพของเรานั้นไปเทรนให้กับโมเดลซึ่งความแม่นยำและความเสถียรนั้นสามารถใช้งานได้ดีและแม่นยำเราจึงเลือกใช้โมเดลนี้

### 3.2 Hyperparameters
Hyperparameter ที่เรามีการเซ็ทค่านั้นคือ epoch หรือจำนวนการเทรนของโมเดลซึ่งเราได้เซ็ทให้เทรนทั้งหมด 1000 ครั้ง ซึ่งเป็นค่าที่เราได้ลองปรับค่าจาก 100 ครั้งซึ่งเมื่อลองนำโมเดลออกมาใช้พบว่าความแม่นยำในการตรวจจับนั้นต่ำมากจนไม่สามารถใช้งานได้ จึงจำเป็นที่จะต้องใช้จำนวนการเทรนขึ้นไปถึง 1000 ครั้ง และมีการกำหนด Batchsize  ของข้อมูลที่จะให้โมเดลเทรนต่อรอบซึ่งเราได้เซ็ทไว้ที่ 32 เพื่อให้จำนวนของข้อมูลต่อรอบที่เทรนนั้นมีเยอะขึ้นแต่ไม่เกิน memory ของเครื่องที่ใช้งานได้
<center><div><img src="img/code.png" height="200"/></div></center>

### 3.3 Model Training

In [ ]:
from ultralytics import YOLO

model = YOLO('yolov8n.pt')

# Path to data.yaml in dataset folder
path = "<Path to data.yaml>"
results = model.train(data=path, epochs=1000, imgsz=640, batch=32)

results = model.val()

success = model.export(format='onnx')

### Result after training
<center><div>
<img src="img/F1_curve.png" height="500"/>
<img src="img/confusion_matrix.png" height="500"/>
</div></center>

# 4. Test

In [2]:
from ultralytics import YOLO
import cv2

model = YOLO("D:/code/train/weights/best.onnx")

# Test from Test image 30%
results = model.predict("D:/Downloads/AI_project.v7i.yolov8/valid/images", save = True, project="Test_img")

WARNING  Unable to automatically guess model task, assuming 'task=detect'. Explicitly define task for your model, i.e. 'task=detect', 'segment', 'classify','pose' or 'obb'.
Loading D:\code\train\weights\best.onnx for ONNX Runtime inference...

image 1/514 D:\Downloads\AI_project.v7i.yolov8\valid\images\-17-758x397_jpg.rf.88f5c71c6c52ddd92b3142883a60bd9f.jpg: 640x640 3 helmets, 3 motorbikes, 113.1ms
image 2/514 D:\Downloads\AI_project.v7i.yolov8\valid\images\-19-_jpg.rf.31a95c86ba2233e9ef89779a217c1592.jpg: 640x640 2 heads, 95.9ms
image 3/514 D:\Downloads\AI_project.v7i.yolov8\valid\images\-20-_31363-19_-_0075-1024x682_jpg.rf.c8c72fe517a36f1206beabf1476598a8.jpg: 640x640 1 head, 2 helmets, 2 motorbikes, 98.4ms
image 4/514 D:\Downloads\AI_project.v7i.yolov8\valid\images\-_jpg.rf.0c29f1d35655e79df15411a199d58b1a.jpg: 640x640 1 helmet, 1 motorbike, 88.4ms
image 5/514 D:\Downloads\AI_project.v7i.yolov8\valid\images\100_jpg.rf.53c6d65de8d418731a098e1c9347e497.jpg: 640x640 1 helmet, 1 motorbi

### Example(ปล. สามารถดูภาพที่ได้จากการ predict ได้จาก folder:Test_img/predict ตาม directory เดิม)
<center><div>
<img src="img/pic1.jpg" height="500"/>
<img src="img/pic2.jpg" height="500"/>
</div></center>

### 4.1 การประเมินผล
จากผลการทดสอบโมเดลโดยใช้ข้อมูลทั้งหมด 30% จะได้ confusion metrics ดังนี้
<div>
<center><img src="img/confusion_matrix.png" height="500" style="display=block; margin:auto"/></center>
</div>

ซึ่งเราสามารถสรุปข้อมูลของ True positive, Fasle positive, False negative, True negative ของแต่ละคลาสได้ดังนี้
<div>
<center><img src="img/table.png" height="200" style="display=block; margin:auto"/></center>
</div>

เราสามารถคำนวนค่า Precision, Recall และ F1-score ได้จากตารางข้างต้นดังนี้
<div>
<center><img src="img/table2.png" height="200" style="display=block; margin:auto"/></center>
</div>

ซึ่งโมเดลของเราจะมีค่า Precision, Recall และ F1-scoreโดยรวมเป็น
<div>
<center><img src="img/table3.png" height="200" style="display=block; margin:auto"/></center>
</div>


In [5]:
from ultralytics import YOLO

# Load YOLOv8 model (nano version or custom weights)
model = YOLO("D:/Downloads/automatic-number-plate-recognition-python-yolov8-main/train12/train12/weights/best.pt")  # or 'path/to/your/custom/weights.pt'

# Validate on the custom dataset and calculate mAP@50
results = model.val(data="D:/Downloads/AI_project.v7i.yolov8/data.yaml", iou=0.5)

Ultralytics YOLOv8.2.103  Python-3.10.11 torch-2.4.1+cpu CPU (Intel Core(TM) i7-9750H 2.60GHz)
Model summary (fused): 168 layers, 3,006,233 parameters, 0 gradients, 8.1 GFLOPs


val: Scanning D:\Downloads\AI_project.v7i.yolov8\valid\labels.cache... 514 images, 0 backgrounds, 0 corrupt: 100%|██████████| 514/514 [00:00<?, ?it/s]
                 Class     Images  Instances      Box(P          R      mAP50  mAP50-95): 100%|██████████| 33/33 [00:57<00:00,  1.74s/it]


                   all        514       1370       0.92      0.941      0.965       0.77
                  head        195        519       0.91      0.948      0.978      0.764
                helmet        236        432      0.918      0.944       0.96       0.75
             motorbike        291        419      0.931      0.931      0.957      0.796
Speed: 1.9ms preprocess, 94.8ms inference, 0.0ms loss, 0.6ms postprocess per image
Results saved to C:\Users\asus\runs\detect\val8


### 4.2 การอธิบายผลลัพธ์
จากผลลัพท์ข้างต้นที่เกิดขึ้นมานั้น
1. Confusion matrix: เป็นตารางที่บ่งบอกถึงสิ่งที่โมเดลทำนายออกมาเทียบกับความเป็นจริงซึ่งจะเห็นได้ว่าในตารางที่โมเดลทำนายออกมานั้นจะมีการทำนายผิดพลาดจากพื้นหลังกลายเป็นหัว หมวกกันน็อค และมอเตอร์ไซต์ซึ่งแสดงให้เห็นว่าโมเดลอาจเกิดข้อผิดพลาดในคานี้ได้เนื่องจากในขั้นตอนการ Label นั้นมีบางวัตถุที่เห็นได้ไม่ชัดเจนจึงไม่ได้ทำการ Label เข้าไปเมื่อโมเดลนำภาพนั้นมาทดสอบอาจจะทำให้ตรวจจับวัตถุนั้นเจอ แต่หากเทียบการทำนายที่ผิดพลาดของ False negative นั้นจะพบว่าในส่วนของหัวและหมวกกันน็อคเกิดความสับสนขึ้นเพียงเล็กน้อยเท่านั้น
2. Precision: คือค่าความแม่นยำที่โมเดลสามารถทำนายได้ว่าเป็นวัตถุชนิดนั้นได้ถูกต้องเป็นเท่าไร โดยที่โมเดลของเรานั้นมีค่า Precision ที่ 81% ซึ่งค่อนข้างสูงแต่ยังไม่เพียงพอ
3. Recall: คือความถูกต้องว่าโมเดลมีโอกาสที่จะทำนายวัตถุนั้นได้ถูกต้องตามความจริงเป็นเท่าไร โดยที่โมเดลของเรานั้นมีค่า Recall ที่ 88% ซึ่งค่อนข้างสูง
4. F1-Score: คือค่าเฉลี่ยของทั้ง Precision และ Recall (คล้าย Accuracy) โดยที่โมเดลของเรานั้นได้ค่าเป็น 84% ซึ่งถือว่าสูงแต่ยังสามารถเพิ่มความแม่นยำได้อีก
5. mAP50: คือการเปรียบเทียบ Overlayว่าการซ้อนทับของ label จริง เทียบกับที่โมเดลทำนายนั้นซ้อนทับกันขนาดไหนซึ่ง โมเดลของเรานั้นมีค่า mAP50 สูงถึง 0.965

# 5. Model Improvement

### 5.1 การวิเคราะห์ข้อบกพร่องของโมเดล
จากการได้ทดลองใช้ในสถานการณ์จริงพบว่าตัวของโมเดลในบางครั้งยังมีข้อผิดพลาดในการตรวจจับและแยกแยะระหว่างหัวกับหมวกกันน็อคโดยเฉพาะเมื่อหมวกน็อคเป็นสีดำซึ่งในบางครั้งที่เป็นหมวกกันน็อคแบบครึ่งใบจะถูกตรวจจับเป็นหัวแทน ซึ่งสามารถปรับปรุงได้โดยการเพิ่มจำนวนของข้อมูลให้มีความหลากหลายมากยิ่งขึ้นเพื่อให้ตรวจจับได้แม่นยำขึ้น

### 5.2 การปรับปรุงโมเดล
โดยเราได้ทำการปรับปรุงโมเดลโดยการเพิ่มจำนวนภาพของหัวและหมวกกันน็อคอีก 200 ภาพคละกันโดยยังใช้การเทรนทั้งหมด 1000 รอบเช่นเดิม
<div>
<center><img src="img/step_imple.png" height="200" style="display=block; margin:auto"/></center>
</div>

### 5.3 การประเมินผลหลังการปรับปรุง
ผลลัพท์จากการเพิ่มจำนวนข้อมูลนั้นพบว่าโมเดลมีความแม่นยำเพิ่มขึ้นจาก 83% เป็น 87% ซึ่งความแม่นยำเพิ่มขึ้นถึง 5% 
|Before|After|
|---|---|
|<img src="img/F1_curve_before.png" height="400" style="display=block; margin:auto"/>|<img src="img/F1_curve.png" height="400" style="display=block; margin:auto"/>|
